proposals/collection-expressions.md

```diff
@@ -112,7 +112,22 @@ Collection literals are [target-typed](https://github.com/dotnet/csharplang/blob
 
     * The behavior of collection literals with collections that are not well-behaved is undefined.
 
+## Element type
+[element-type]: #element-type
+
+Certain uses of *collection expressions* &mdash; [*type inference*](#type-inference) and [*create methods*](#create-methods) &mdash; require a target type with a known *element type*.
+
+The *element type* of a type is defined as:
+
+* For a single dimensional *array type* `T[]`, the *element type* is `T`.
+* For a *span type* `System.Span<T>` or `System.ReadOnlySpan<T>`, the *element type* is `T`.
+* For an *[inline array type](https://github.com/dotnet/csharplang/blob/main/proposals/inline-arrays.md)* with *element type* `T`, the *element type* is `T`.
+* For a *type* that implements `System.Collections.IEnumerable`, the *element type* is `System.Object`.
+* For an *interface type* `I<T0>` where `System.Collections.Generic.List<T>` implements `I<T>`, the *element type* is `T0`.
+* Otherwise the type does not have an *element type*.
```

Comment:
i'm actually surrpised it's defined this way.  For example, the above seems like it would mean that a type that is `foreachable` but not any of the above, would not have an element type?   my understanding is that foreach operates directly as a pattern, looking for a `GetEnumerator()` with a certain shape, which returns an type with a certain shape.  I would expect it would the be the type of `X Current { get; }` that would define the element type.

Reply:
basically, i woudl expect the element type to link to the algorithm here: https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement (or some variation thereof).  Basically:

> The compile-time processing of a foreach statement first determines the collection type, enumerator type and iteration type of the expression. This determination proceeds as follows:

Here, the `element type` seems to equivalent to the `iteration type` specified in this other part of the spec (modulo if we want to make any tweaks).  But it seems to serve better than writing out this different set of rules (which may not be as inclusive, or may deviate in unintended ways?).

Also, wrt the impl, are we sharing the code to figure this out with the foreach code?  If so, then unifying in the spec seems very preferable.

Reply:
I overlooked the `GetEnumerator()` pattern here. As you suggested, if the type implements the `GetEnumerator()` pattern, we should the `Current` property for the element type.

That said, I think the definition of *element type* might differ from *`foreach` iteration type* in two ways: extensions and accessibility.

In short, the element type should not depend on where the *collection expression* is used. Instead, we should be able to determine the element type at the definition of the target type rather than at the use-site. If so, that might mean the `GetEnumerator()` implementation should be a `public` instance method, and should not depend on extension methods or accessibility.

Reply:
> and should not depend on extension methods or accessibility

So I personally disagree. Since it feels like anything foreachable should reasonably both be able to contribute one of these element types, and then work in an these collection-expr scenarios.

Reply:
Allowing extension method `GetEnumerator()` may lead to surprising behavior when the extension method is not in scope or inaccessible.

For instance, in the following example, a `MyCollection<T>` is constructed successfully in one case, but not the other.

Should we report a diagnostic (perhaps a warning) at the declaration of `class MyCollection<T>` when there is no instance `GetEnumerator()` that use-sites may not be able to construct the collection from a collection expression?

```csharp
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

[CollectionBuilder(typeof(MyCollectionBuilder), nameof(MyCollectionBuilder.Create))]
class MyCollection<T> { }

class MyCollectionBuilder
{
    public static MyCollection<T> Create<T>(ReadOnlySpan<T> items) => default;
}

namespace N
{
    static class Extensions
    {
        public static IEnumerator<T> GetEnumerator<T>(this MyCollection<T> c) => default;
        static MyCollection<T> F<T>() => []; // ok
    }
}

class Program
{
    static void Main()
    {
        MyCollection<int> c = []; // error: CollectionBuilderAttribute but no element type
    }
}
```

Reply:
I don't think it's any more surprising than being able to foreach or not over instances based on it the extension is in scope :-)

Reply:
Stuff you can add with just extensions:
- ability to 'foreach' (extension GetEnumerator)

Stuff you can add with both declaration changes+extensions:
- ability to use collection initializers. Declaration needs to implement IEnumerable but an extension method can implement Add.

Stuff you can add only with declaration changes
- ability to use a `Create(ROS<T>)` method for a collection expression. The declaration needs to also add CollectionBuilderAttribute. Extension GetEnumerator doesn't allow us to determine the element type. It needs to be declared on the collection type.

On a certain level, I feel like we've decided to just focus on having original declarations onboard to this pattern, and I feel totally fine with that, as long as we have some idea of how they could be added through extensions in the future, which I think we do.

Reply:
> On a certain level, I feel like we've decided to just focus on having original declarations onboard to this pattern, 

I don't recall ever makgin this decision.  And, personally, it goes against the spirit of Collection Exprs.  The point of collection exprs is to be encompassing.  If people have their own collections then they should work naturally with collection exprs.

We're not creating a new, opinionated, system that intends to leave things behind.